src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs

```diff
@@ -65,6 +65,56 @@ void M3()
                 );
         }
 
+        [Fact]
+        public void TestMissingGreaterThanToken()
+        {
+            var text = @"
+            class Class1
+            {
+                public static IEnumerable<(String Value, String Description) GetAllValues(Type t)
+                {
+                    if (!t.IsEnum)
+                        throw new ArgumentException(""no good"");
+
+                    return Enum.GetValues(t).Cast<Enum>().Select(e => (e.ToString(), e.ToString()));
+                }
+            }";
+
+            var file = this.ParseTree(text);
+
+            Assert.NotNull(file);
+            Assert.Equal(text, file.ToFullString());
+            file.Errors().Verify(
+                // error CS1003: Syntax error, '>' expected
+                Diagnostic(ErrorCode.ERR_SyntaxError).WithArguments(">", "")
+            );
+        }
+
+        [Fact]
+        public void TestMissingCommaAndGreaterThanToken()
```

Comment:
There are tests that validate the shape of teh actual tree created.  Look at UsingTree and the tests that use that.  Thanks! :)

Reply:
Added tests for checking actual tree created. Should I remove these tests, which check error messages, or leave them here?

Reply:
Generally speaking, a syntax test should assert both the diagnostics and the shape of the resulting tree.

---
In reply to: [169297996](https://github.com/dotnet/roslyn/pull/24911#discussion_r169297996) [](ancestors = 169297996)